src/grammar.coffee

```diff
@@ -392,11 +392,11 @@ grammar =
   # or by array index or slice.
   Accessor: [
     o '.  Property',                            -> new Access $2
-    o '?. Property',                            -> new Access $2, 'soak'
-    o ':: Property',                            -> [LOC(1)(new Access new PropertyName('prototype')), LOC(2)(new Access $2)]
-    o '?:: Property',                           -> [LOC(1)(new Access new PropertyName('prototype'), 'soak'), LOC(2)(new Access $2)]
-    o '::',                                     -> new Access new PropertyName 'prototype'
-    o '?::',                                    -> new Access new PropertyName('prototype'), 'soak'
+    o '?. Property',                            -> new Access $2, soak: yes
+    o ':: Property',                            -> [LOC(1)(new Access new PropertyName('prototype'), shorthand: yes), LOC(2)(new Access $2)]
```

Comment:
Tell the `Access` whether it was `shorthand` (ie `::`) or not so that it can be exposed on the AST

Reply:
Sounds good. Probably not in this PR, but we need to think about the style we’re using for these inline objects in `grammar.coffee`. I try to avoid seemingly ambiguous cases like `fn a, b: c, d: e`. Reading it carefully, it’s clear that it needs to be `fn a, {b: c, d: e}` but it’s easy for a quick reader to not be aware of where the object starts and ends. I don’t know what a better style is; just adding the braces? Making all such objects indented on new lines? I’m not sure, but I don’t like how the current style can be confusing. (And I know you didn’t come up with this, that it was this way before.)

Reply:
The issue is that the object often isn’t the only argument of the function, or even the last argument: `fn arg1, obj, arg2`. I agree, the “options argument” pattern is way more flexible than dealing with sequential function arguments, and I would be in favor if refactoring the codebase to standardize on that potentially; but in its own PR 😄For now I’m trying to just follow the style of the functions we have, and not _mix_ styles. I’m hoping that once all this Prettier stuff is done we do a big cleanup pass for style, assisted by Prettier reformatting 😄

Reply:
Ya to be clear, I'm saying that my tendency is to only omit object braces if the object is trailing, so a convention where `fn arg1, arg2, obj` gets formatted `fn arg1, arg2, a: ..., b: ...` but `fn arg1, obj, arg2` gets formatted `fn arg1, {a: ..., b: ...}, arg2`

If you follow that convention, then you can assume that you're seeing a single trailing object argument if you see an implicit object amongst call args, which makes it significantly visually easier to parse (than if you had to scan whether it was in fact a trailing arg, a single object, ...)

That convention is like Ruby in the sense that (from my understanding, maybe this was only pre-Ruby 2.0) Ruby only allows you to use a syntax that looks like implicit object args if it's a single trailing object (hash)

Speaking of Prettier, that's actually the current default behavior of the Prettier formatter is that it'll put explicit braces around a non-trailing object arg but leave it implicit if it's trailing